docs/01-development/design-docs/003-stateless-redis-disruptor.md

```diff
@@ -0,0 +1,117 @@
+
+RESP is pipelined, which means that the same connection may be used for the client to send multiple requests, without waiting for the server to respond to each one. This means that for a proxy to be able to know to which request a response belongs, it needs to keep in memory the request that originated it. Such a proxy is referred in this document as a stateful proxy, as it needs to keep in memory the state of the command queue, and introduces complexity to the proxy.
+
+Without the requirement of being able to correlate responses with the requests that originated them, a RESP proxy can be made stateless. This reduces the complexity at the cost of, as expected, not being to correlate those responses. However, it should still be possible to meet the goals above with an stateless proxy.
+
+A stateless RESP proxy accepts connections from Redis clients. It will read messages sent by clients, parse them, and decide if any action is necessary, such as modifying the request, or delaying it. It simply passes through responses from the server back to the client, without needing to decode them. A stateless proxy always needs to forward requests, modified or not, to the upstream server. As it is not aware of the flow of responses, it should be compatible with server pushes without needing any additional logic. 
+
+A stateless RESP proxy can be used as a first step to meet the goals above, with some limitations.
+
+### Stateless delay
+
+For each message received by the RESP proxy proxy, it will wait a certain amount of time before forwarding it upstream. To preserve the protocol's semantics, latency would likely need to be added per-message, even if it includes multiple commands.
+
+As the stateless proxy does not match requests to responses, latency will always be added to the upstream server latency.
+
+### Stateless miss rate
+
+For each command that retrieves one or more keys, for each key that matches a user-specified prefix a random number will be generated and compared to a user-specified threshold. If the number is smaller than the threshold, the key is modified to a randomly generated value that will most likely not exist.
+
+For example, the if the following commands arrive on a batch:
+```
+GET users:1234
+GET users:1235
+GET users:1236
+```
+
+And we are faking a 33% miss rate, the proxy would modify that batch and send the following upstream:
+```
+GET users:1234
+GET __xk6_disrupted_1693995714__
+GET users:1236
+```
+
+#### Caveats
+
+A stateless proxy cannot inject fully consistent miss rate faults, as it wouldn't be able to affect commands used to list the keys present in Redis such as:
+- `KEYS`
+- `RANDOMKEY`
+- `SCAN`
+
+This needs to be acknowledged as a limitation of the stateless approach.
+
+### Advantages
+
+- Easier to implement and less error-prone than a stateful proxy
```

Comment:
I think that it is valid to implement a PoC using the stateless approach, but I would prefer that we address a full-fledged implementation in this design document.